WIP sql: support output parameters in procedures

yuzefovich opened the pull request:
See each commit for details.

Addresses: #100405.
Epic: CRDB-30611

yuzefovich committed:
sql: support OUT parameters for overload resolution in procedures

This commit finalizes the overload resolution for procedures. The main
difference between procedures and UDFs is that for procedures all
parameters need to be specified in the CALL invocation, including OUT
parameters. Thus, the original idea for implementing the overload
resolution (which was already partially achieved in the code) was to
include all parameters into the "signature" (unlike for UDFs where only
input parameters are included). This turned out to be tricky to
implement because in some contexts (e.g. in DROP or REPLACE) OUT
parameters need to be ignored.

This commit rips out this signature customization for procedures to only
include the input parameters. However, in order to ignore the OUT
parameters during the type checking (to match the correct overloads)
this required additional plumbing to indicate whether a particular
argument is for the OUT parameter. We need to include this information
into the FunctionSignature proto which is done in backwards-compatible
way, and the necessary type-checking code that looks at the types of the
parameters has been adjusted accordingly.

An additional quirk of procedures when comparing to UDFs is that if any
output parameters are specified, they define a RECORD return type (even
with a single column). This required some adjustments to the existing
logic.

Note that this commit doesn't have any tests that actually call the
procedures, because that logic is implemented in the following commit
(previously, since we didn't support output parameters, procedures
always returned VOID / NULL).

Release note: None

yuzefovich committed:
sql: support returning results from procedures with output parameters

This commit teaches the execution to return a single row from procedures
with output parameters. Whenever a procedure has any output parameters,
it's expected to return a single tuple elements from which form the
result row of the procedure call. Additionally, this commit adjusts the
logic to skip evaluation of OUT argument expressions (to match
postgres) and ignore them from passing into the udf call.

It also extends the logic test framework a bit to add a mode that
asserts that the output of a query is empty (procedures with no output
parameters don't return any rows).

Release note (sql change): OUT and INOUT parameter classes are now
supported in stored procedures.

yuzefovich committed:
TODO: squash in previous commit

This commit updates the previous commit in the following ways:
- we now include only OUT parameter ordinals into the FunctionSignature
proto (rather than "to input ordinal mapping")
- we now also include types of all OUT parameters
- we populate this information for UDFs too - this will be utilized in
a separate PR to match Postgres in some scenarios
- type-checking logic for CALL expressions has been extended so that if
it doesn't find the procedure to invoke, it then attempts to treat UDFs
as procedures so that a more useful error message is returned (i.e.
instead of "procedure does not exist" we now will get "is not
a procedure" - matching postgres)
- we now perform type-checking of expressions passes for OUT parameters
in CALL invocations - according to PG docs, these expressions must be
coercible to the types from procedure definition, but the expressions
themselves are not evaluated.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/120731)
<!-- Reviewable:end -->

yuzefovich committed:
tree: introduce "default" parameter class when left unspecified

This commit introduces `RoutineParamDefault` which is used whenever the
parameter class is left unspecified. In almost all cases it is
equivalent to `RoutineParamIn` (i.e the `IN` class), but in DROP
PROCEDURE it can have a different behavior, so we need to be able to
distinguish between `ParamIn` and `ParamDefault`. This required a minor
change to the parser. Note that we already had "unspecified" param class
proto value, so this change should be backwards-compatible.

Many tests are now adjusted to better represent reality (i.e. have
`(param INT) instead of `(IN param INT)` when the former was actually
used on creation).

Release note: None

yuzefovich committed:
sql: implement SQL-compliant DROP PROCEDURE

This commit extends the overload resolution logic to match the
SQL-compliant behavior of PG on DROP PROCEDURE when no parameter has the
parameter class specified (i.e. all parameters use the "default" class).
In such a scenario, when matching the requested signature to the
overload PG applies two matching approaches:
- the main one - based on the input types
- the SQL-compliant one where all parameters are considered.

For context, SQL standard allows function and procedure overloads only
when they have different number of parameters (regardless of their
class), but PG allows more - namely the input parameters must be
different between overloads. To partially bring the PG behavior closer
to the spec they implemented this two-way resolution approach in
https://github.com/postgres/postgres/commit/e56bce5d43789cce95d099554ae9593ada92b3b7
which can lead to ambiguity errors in some rare cases.

Release note: None
